src/contexts/partner/PartnerContext.tsx

```diff
@@ -0,0 +1,164 @@
+  alertToDisplay: undefined,
+  redirectUrl: '',
+})
+
+export function PartnerProvider({ children }: PropsWithChildren) {
+  const [alertToDisplay, setAlertToDisplay] = useState<
+    AlertToDisplay | undefined
+  >(undefined)
+  const [redirectUrl, setRedirectUrl] = useState('')
+
+  const searchParams = useSearchParams()
+
+  const { progression, situation } = useCurrentSimulation()
+
+  const { exportSituationAsync } = useExportSituation()
+
+  const router = useRouter()
+
+  const partnerParams: Record<string, string> | undefined = useMemo(() => {
+    const params =
+      getPartnerFromStorage() ??
+      Object.fromEntries(
+        searchParams
+          .entries()
+          .filter(([key]) => key === PARTNER_KEY || key.startsWith(PARTNER_KEY))
+      )
+
+    return Object.keys(params).length ? params : undefined
+  }, [searchParams])
+
+  const isPartnerVerified = useVerifyPartner(partnerParams?.partner)
+
+  const hasNoPartnerParam = Object.keys(partnerParams || {}).length === 0
+
+  const handleExportSituation = useCallback(async () => {
+    if (!partnerParams) return
+
+    try {
+      const { redirectUrl: redirectUrlFromResponse } =
+        (await exportSituationAsync({
+          situation,
+          partner: partnerParams[PARTNER_KEY],
+          partnerParams,
+        })) ?? {}
+
+      if (!redirectUrlFromResponse) throw new Error('No redirect URL')
+
+      setRedirectUrl(redirectUrlFromResponse)
+
+      setAlertToDisplay({
+        type: 'success',
+        content: <SuccessMessage redirectUrl={redirectUrlFromResponse} />,
+      })
+    } catch (error) {
+      captureException(error)
+
+      setAlertToDisplay({
+        type: 'error',
+        content: (
+          <div className="xs:text-left text-center">
+            <span className="mb-2 block">
+              <Emoji className="mr-2 text-base">😫</Emoji>
+              <Trans>
+                Oups ! Une erreur s'est produite au moment d'envoyer vos
+                résultats à notre partenaire. Veuillez recommencer le parcours
+                depuis l'interface du site partenaire.
+              </Trans>
+            </span>
+          </div>
+        ),
+      })
+    } finally {
+      removePartnerFromStorage()
+    }
+    // False positive
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [exportSituationAsync, situation, partnerParams, router])
+
+  useEffect(() => {
+    // Redirect to 404
+    if (
+      partnerParams &&
+      typeof isPartnerVerified !== 'undefined' &&
+      !isPartnerVerified
+    ) {
+      router.push('/404')
+    }
+  }, [isPartnerVerified, router, partnerParams])
+
+  useEffect(() => {
+    if (
+      !isPartnerVerified ||
+      hasNoPartnerParam ||
+      typeof progression === 'undefined' ||
+      !partnerParams
+    )
+      return
+
+    if (progression === 1) {
+      handleExportSituation()
+    } else if (progression !== 1 && !getPartnerFromStorage()) {
```

Comment:
Maybe use `partnerParams` to not lookup `sessionStorage` again ?

Reply:
Ici on veut vérifier que le session storage ne contient pas `partnerParams` ; si je vérifie que `partnerParams` est vide, on ne les stockera jamais